server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -100,73 +103,92 @@ class ObjectService(
-                val localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath)
-                for {
-                  fileBody <- io.file.readAll[IO](localAbsolutePath, blockingEc, 4096).compile.to[Array]
-                  calculatedCrc32c = Crc32c.calculateCrc32c(fileBody)
-                  syncStatus <- if (calculatedCrc32c == crc32c) IO.pure(SyncStatus.Live) else {
-                    for {
-                      cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath))
-                      status <- cachedGeneration match {
-                        case Some(cachedGen) => if(cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
-                        case None => IO.pure(SyncStatus.OutOfSync)
-                      }
-                    } yield status
-                  }
-                  currentTime <- timer.clock.realTime(TimeUnit.MILLISECONDS)
-                  lastLock <- expiresAt.flatTraverse[IO, WorkbenchEmail] { ea =>
-                    if (currentTime > ea.toEpochMilli)
-                      IO.pure(none[WorkbenchEmail])
-                    else
-                      IO.pure(lastLockedBy)
-                  }
-                } yield MetadataResponse.EditMode(syncStatus, lastLock, generation, sl)
-            }
-          } yield result
-        }
+  def checkMetadata(req: GetMetadataRequest): Kleisli[IO, TraceId, MetadataResponse] = {
+    findStorageLink(
+      req.localObjectPath,
+      pair => Kleisli.liftF(IO.pure(MetadataResponse.SafeMode(pair.storageLink))),
+      pair => Kleisli { traceId =>
+        val fullBlobName = getFullBlobName(pair.basePath, req.localObjectPath, pair.storageLink.cloudStorageDirectory.blobPath)
+        for {
+          metadata <- retrieveGcsMetadata(req.localObjectPath, pair.storageLink.cloudStorageDirectory.bucketName, fullBlobName, traceId)
+          result <- metadata match {
+            case None =>
+              IO(MetadataResponse.RemoteNotFound(pair.storageLink))
+            case Some(GcsMetadata(_, lastLockedBy, expiresAt, crc32c, generation)) =>
+              val localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath)
+              for {
+                fileBody <- io.file.readAll[IO](localAbsolutePath, blockingEc, 4096).compile.to[Array]
+                calculatedCrc32c = Crc32c.calculateCrc32c(fileBody)
+                syncStatus <- if (calculatedCrc32c == crc32c) IO.pure(SyncStatus.Live) else {
+                  for {
+                    cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath))
+                    status <- cachedGeneration match {
+                      case Some(cachedGen) => if(cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
+                      case None => IO.pure(SyncStatus.OutOfSync)
```

Comment:
What is OutOfSync? Is this the same as Desynchronized? If so - why a different name?